SDK support for implicit namespaces in C# projects

## SDK support for implicit namespaces in C# projects

We are introducing implicit namespace support for C# projects in the .NET SDK. The goal is to reduce the amount of boilerplate in .NET C# project templates by removing the need for `using` statements at the top of most files with the [global using feature](https://github.com/dotnet/csharplang/issues/3428) introduced in C# 10. 

Implicit namespace will be enabled by default for C# projects targeting the `net6.0` TFM or higher and using `Microsoft.NET.Sdk`, `Microsoft.NET.Sdk.Web` or `Microsoft.NET.Sdk.Worker` SDK. A generated file containing the default namespaces will be included in the set of files passed to the compiler. Projects using a C# version less than 10 need to explicitly disable this feature.

### Modifying the list of implicit namespaces 

There are three ways to control the namespaces that are included in the project:

1. Disable the feature completely by setting `<DisableImplicitNamespaceImports>` to `true` in the project file. 
2. Disable a set of implicit namespaces added by a SDK, see the list of SDK specific namespace imports below for the setting name.
3. Add or remove individual namespaces in the project file by modifying the `<Import>` item group in the project file. For example

```xml
<ItemGroup>
  <Import Remove="System.Net.Http" />
  <Import Include="System.IO.Pipes" />
</ItemGroup>
```

These are the implicit namespaces added to projects using the following SDKs:

### Microsoft.NET.Sdk 

The following list of namespaces will be added:

```
System
System.Collections.Generic
System.IO
System.Linq
System.Net.Http
System.Threading
System.Threading.Tasks
```

This set of implicit namespace imports can be disabled by setting `<DisableImplicitNamespaceImports_DotNet>` to `true` in the project file. 

### Microsoft.NET.Sdk.Web 

The following list of namespaces will be added in addition to the ones added by the Microsoft.NET.Sdk:

```
System.Net.Http.Json
Microsoft.AspNetCore.Builder
Microsoft.AspNetCore.Hosting
Microsoft.AspNetCore.Http
Microsoft.AspNetCore.Routing
Microsoft.Extensions.Configuration
Microsoft.Extensions.DependencyInjection
Microsoft.Extensions.Hosting
Microsoft.Extensions.Logging
```

This set of implicit namespace imports can be disabled by setting `<DisableImplicitNamespaceImports_Web>` to `true` in the project file. 

### Microsoft.NET.Sdk.Worker

The following list of namespaces will be added in addition to the ones added by the Microsoft.NET.Sdk:

```
Microsoft.Extensions.Configuration
Microsoft.Extensions.DependencyInjection
Microsoft.Extensions.Hosting
Microsoft.Extensions.Logging
```

This set of implicit namespace imports can be disabled by setting `<DisableImplicitNamespaceImports_Worker>` to `true` in the project file. 

### Version introduced

6.0-preview7

### Old behavior

No implicit namespaces are added for applicable C# projects.

### New behavior

Implicit namespaces are added via global usings in a generated file in the obj directory. See [how to control the list of included implicit namespaces](#modifying-the-list-of-implicit-namespaces) above.

### Category

<!-- Choose one of the following compatibility categories that this change affects: -->

- [ ] Binary compatibility (code must be recompiled to use the newer API version)
- [x] Source compatibility (successfully recompiling against the newer API version requires source changes)

There is a potential need to resolve type conflicts due to the introduction of implicit namespaces. 

### Reason for change

These changes in the SDK are made to reduce the amount of boilerplate `using` statements that appear at the top of most C# projects.

### Recommended action

This feature will be enabled by default and for most users, no action will need to be taken. However, the introduction of this mechanism can cause type name conflicts with the namespaces listed above. In those cases, action need to be take to [modify the list of implicit namespaces](#modifying-the-list-of-implicit-namespaces)

### Feature area

<!-- Choose an area from one of the following: -->

- [x] ASP.NET Core
- [x] C#
- [ ] Code analysis
- [ ] Core .NET libraries
- [ ] Cryptography
- [ ] Data
- [ ] Debugger
- [ ] Deployment
- [ ] Globalization
- [ ] Interop
- [ ] JIT
- [ ] LINQ
- [ ] Managed Extensibility Framework (MEF)
- [ ] MSBuild
- [ ] Networking
- [ ] Printing
- [x] SDK
- [ ] Security
- [ ] Serialization
- [ ] Visual Basic
- [ ] Windows Forms
- [ ] Windows Presentation Foundation (WPF)
- [ ] XML, XSLT

### Affected APIs



Finally, please remember to email a link to this breaking change issue to [.NET Breaking Change Notifications](mailto:dotnetbcn@microsoft.com).

<!-- Do not modify anything below this line -->

---
#### Issue metadata

* Issue type: breaking-change

It might be helpful to show one potential error you might get, specifically if you're authoring an MSBuild task, and thus deriving from the Task abstract class, you will see errors like this because it conflicts with the Task class in System.Threading.Tasks, which is now being implicitly included:

```
...: error CS0104: 'Task' is an ambiguous reference between 'Microsoft.Build.Utilities.Task' and 'System.Threading.Tasks.Task' 
.... error CS0115: 'SomeTask.Execute()': no suitable method found to override 
```

If you get the above, you will need to disable this feature for the project.

--- a/docs/core/compatibility/sdk/6.0/implicit-namespaces.md
+++ b/docs/core/compatibility/sdk/6.0/implicit-namespaces.md
@@ -17,7 +17,7 @@ If your project uses a C# version less than 10, you must explicitly [disable thi
 
 ## Old behavior
 
-No implicit namespaces are added to C# projects.
+No implicit namespaces are added to C# projects and there are no type conflicts.
 
 ## New behavior
 
@@ -31,9 +31,16 @@ The default namespaces are included by adding `global using` directives to a gen
 
 | SDK | Default namespaces |
 | - | - |
-| Microsoft.NET.Sdk | System<br/>System.Collections.Generic<br/>System.IO<br/>System.Linq<br/>System.Net.Http<br/>System.Threading<br/>System.Threading.Tasks |
-| Microsoft.NET.Sdk.Web | System.Net.Http.Json<br/>Microsoft.AspNetCore.Builder<br/>Microsoft.AspNetCore.Hosting<br/>Microsoft.AspNetCore.Http<br/>Microsoft.AspNetCore.Routing<br/>Microsoft.Extensions.Configuration<br/>Microsoft.Extensions.DependencyInjection<br/>Microsoft.Extensions.Hosting<br/>Microsoft.Extensions.Logging |
-| Microsoft.NET.Sdk.Worker | Microsoft.Extensions.Configuration<br/>Microsoft.Extensions.DependencyInjection<br/>Microsoft.Extensions.Hosting<br/>Microsoft.Extensions.Logging |
+| Microsoft.NET.Sdk | <xref:System><br/><xref:System.Collections.Generic?displayProperty=fullName><br/><xref:System.IO?displayProperty=fullName><br/><xref:System.Linq?displayProperty=fullName><br/><xref:System.Net.Http?displayProperty=fullName><br/><xref:System.Threading?displayProperty=fullName><br/><xref:System.Threading.Tasks?displayProperty=fullName> |
+| Microsoft.NET.Sdk.Web | <xref:System.Net.Http.Json?displayProperty=fullName><br/><xref:Microsoft.AspNetCore.Builder?displayProperty=fullName><br/><xref:Microsoft.AspNetCore.Hosting?displayProperty=fullName><br/><xref:Microsoft.AspNetCore.Http?displayProperty=fullName><br/><xref:Microsoft.AspNetCore.Routing?displayProperty=fullName><br/><xref:Microsoft.Extensions.Configuration?displayProperty=fullName><br/><xref:Microsoft.Extensions.DependencyInjection?displayProperty=fullName><br/><xref:Microsoft.Extensions.Hosting?displayProperty=fullName><br/><xref:Microsoft.Extensions.Logging?displayProperty=fullName> |
+| Microsoft.NET.Sdk.Worker | <xref:Microsoft.Extensions.Configuration?displayProperty=fullName><br/><xref:Microsoft.Extensions.DependencyInjection?displayProperty=fullName><br/><xref:Microsoft.Extensions.Hosting?displayProperty=fullName><br/><xref:Microsoft.Extensions.Logging?displayProperty=fullName> |
+
+These implicit namespaces may cause type conflicts. For example, if you're authoring an MSBuild task, and thus deriving from the abstract <xref:Microsoft.Build.Utilities.Task?displayProperty=fullName> class, you'll see a conflict with the <xref:System.Threading.Tasks.Task?displayProperty=fullName> class. This conflict occurs because the <xref:System.Threading.Tasks?displayProperty=fullName> namespace is implicitly included. The error looks similar to:
+
+```txt
+... error CS0104: 'Task' is an ambiguous reference between 'Microsoft.Build.Utilities.Task' and 'System.Threading.Tasks.Task' 
+... error CS0115: 'SomeTask.Execute()': no suitable method found to override
+```
 
 ## Change category
 

